### PR TITLE
[SPARK-37932][SQL]Wait to resolve missing attributes before applying DeduplicateRelations

### PR DESCRIPTION
**Pull request by chenzhx**
### What changes were proposed in this pull request?

When the join with duplicate view like
```
SELECT l1.idFROM v1 l1
 INNER JOIN (
   SELECT id FROM v1
   GROUP BY id  HAVING COUNT(DISTINCT name) > 1
 ) l2
 ON l1.id = l2.id
 GROUP BY l1.name, l1.id;
```
The error stack is:
```
Resolved attribute(s) name#26 missing from id#31,name#32 in operator !Aggregate [id#31], [id#31, count(distinct name#26) AS count(distinct name#26)#33L]. Attribute(s) with the same name appear in the operation: name. Please check if the right attribute(s) are used.;
Aggregate [name#26, id#25], [id#25]
+- Join Inner, (id#25 = id#31)
   :- SubqueryAlias l1
   :  +- SubqueryAlias spark_catalog.default.v1
   :     +- View (`default`.`v1`, [id#25,name#26])
   :        +- Project [cast(id#20 as int) AS id#25, cast(name#21 as string) AS name#26]
   :           +- Project [id#20, name#21]
   :              +- SubqueryAlias spark_catalog.default.t
   :                 +- Relation default.t[id#20,name#21] parquet
   +- SubqueryAlias l2
      +- Project [id#31]
         +- Filter (count(distinct name#26)#33L > cast(1 as bigint))
            +- !Aggregate [id#31], [id#31, count(distinct name#26) AS count(distinct name#26)#33L]
               +- SubqueryAlias spark_catalog.default.v1
                  +- View (`default`.`v1`, [id#31,name#32])
                     +- Project [cast(id#27 as int) AS id#31, cast(name#28 as string) AS name#32]
                        +- Project [id#27, name#28]
                           +- SubqueryAlias spark_catalog.default.t
                              +- Relation default.t[id#27,name#28] parquet
```
Spark will consider the two views to be duplicates, which will cause the query to fail. 

### Why are the changes needed?

Fix bug when using join in duplicate views.

### Does this PR introduce _any_ user-facing change?

Yes. When we join with duplicate view, the query would be successful.
DeduplicateRelations should only kick in if the plan's children are all resolved and valid.

### How was this patch tested?

Add new UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 1]
```diff
@@ -40,43 +40,49 @@ case class ReferenceEqualPlanWrapper(plan: LogicalPlan) {
...
-            case child =>
-              val projectList = child.output.map { attr =>
-                Alias(attr, attr.name)()
-              }
-              Project(projectList, child)
+    val newPlan = renewDuplicatedRelations(mutable.HashSet.empty, plan)._1
+    if (newPlan.find(p => p.resolved && p.missingInput.nonEmpty).isDefined) {
+      // Wait for `ResolveMissingReferences` to resolve missing attributes first
+      newPlan
+    } else {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:47]: Please remove `else {`.

**Review by beliefer — COMMENTED** [review]
LGTM except one comment.

**Comment by beliefer** [comment]
ping @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
Does Spark 3.1 have this bug?

**Comment by beliefer** [comment]
> Does Spark 3.1 have this bug?

@chenzhx Please investigate the behavior in Spark 3.1.

**Comment by chenzhx** [comment]
This problem is not present on Spark 3.1.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Comment by martinf-moodys** [comment]
Hi @cloud-fan, @chenzhx,
Checking for missing references is very costly on my workflow, due to the call to `missingInput`. Do you think it could be optimized, either by doing this check faster or changing when `DeduplicateRelations` is called so that this check is no longer necessary?

**Comment by peter-toth** [comment]
@martinf-moodys, [SPARK-47319](https://issues.apache.org/jira/browse/SPARK-47319) / https://github.com/apache/spark/pull/45424 might help, especially if you have many `Union` nodes

**Pull request closed without merging**